### PR TITLE
feat: pmtiles worker code re-write

### PR DESCRIPTION
**Pull request by zackpollard**
First part of the code re-write. Next stage is automated deployment of this code as well as auto-update of the map pmtiles file in R2.

### CODE REVIEW

File: tiles/src/repository.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    return stream ?? undefined;
+  }
+}
+
+/* eslint-disable no-var */
+declare global {
+  var memCache: Map<string, unknown>;
+}
+/* eslint-enable no-var */
```
Comment (jrasm91) [inline tiles/src/repository.ts:36]: this _probably_ can go in a `types.d.ts` file

File: tiles/src/repository.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+if (!globalThis.memCache) {
+  globalThis.memCache = new Map<string, unknown>();
+}
```
Comment (jrasm91) [inline tiles/src/repository.ts:40]: this code probably belongs where memCache is used (index.ts)

File: tiles/src/repository.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  ) {}
+
+  getFileName(): string {
+    return this.fileName;
+  }
+
+  private async getR2Object(range: { offset: number; length: number }) {
+    const { offset, length } = range;
+    const resp = await this.bucket.get(this.fileName, {
+      range: { offset: offset, length: length },
```
Comment (jrasm91) [inline tiles/src/repository.ts:66]: { offset, length }

File: tiles/test/globalSetup.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import express from 'express';
+import { AddressInfo } from 'node:net';
+import * as Path from 'node:path';
```
Comment (jrasm91) [inline tiles/test/globalSetup.ts:3]: why is this capitalized lol

File: tiles/test/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      expect(Buffer.from(await response.arrayBuffer()).toString()).toMatchFileSnapshot('./__snapshots__/v1-1-0-1.mvt');
+    });
+    it('responds with error when tile out of bounds', async () => {
+      const response = await SELF.fetch('https://example.com/v1/0/0/1.mvt');
+      expect(response.status).toBe(500);
+    });
+  });
+});
+
+describe.skip('Hello World worker', () => {
```
Comment (jrasm91) [inline tiles/test/index.spec.ts:39]: do these still not work/need to be skipped?

File: tiles/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+		"checkJs": false /* Enable error reporting in type-checked JavaScript files. */,
+		"noEmit": true /* Disable emitting files from a compilation. */,
+		"isolatedModules": true /* Ensure that each file can be safely transpiled without relying on other imports. */,
+		"allowSyntheticDefaultImports": true /* Allow 'import x from y' when a module doesn't have a default export. */,
+		"forceConsistentCasingInFileNames": true /* Ensure that casing is correct in imports. */,
+		"strict": true /* Enable all strict type-checking options. */,
+		"skipLibCheck": true /* Skip type checking all .d.ts files. */
+	},
+	"exclude": ["test"],
+	"include": ["worker-configuration.d.ts", "src/**/*.ts"]
```
Comment (jrasm91) [inline tiles/tsconfig.json:22]: you can include any *.d.ts files here, including one to extend globalThis

**Review by jrasm91 — APPROVED** [review]

File: tiles/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  requestType: 'json';
+  version: string;
+};
+
+export function parseUrl(request: Request): PMTilesParams {
+  const url = new URL(request.url);
+  const version = URL_MATCHER.exec(url.pathname)?.groups?.VERSION;
+  const z = URL_MATCHER.exec(url.pathname)?.groups?.Z;
+  const x = URL_MATCHER.exec(url.pathname)?.groups?.X;
+  const y = URL_MATCHER.exec(url.pathname)?.groups?.Y;
```
[the remark this answers is not in the document]
Comment (zackpollard) [reply tiles/src/index.ts:30]: Lol, yes. This was probably me not sanity checking Copilot...

File: tiles/test/index.spec.ts (hunk 5 above)
Comment (zackpollard) [reply tiles/test/index.spec.ts:39]: Not added those unit tests yet

File: tiles/src/env.d.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+interface WorkerEnv
+  extends Omit<Env, 'ENVIRONMENT' | 'PMTILES_FILE_NAME' | 'PMTILES_FILE_HASH' | 'VMMETRICS_API_TOKEN'> {
```
[the remark this answers is not in the document]
Comment (zackpollard) [reply tiles/src/env.d.ts]: I'm not going to do this for now, we shouldn't really need this in a cloud solution. Also if we need to migrate later it's incredibly simple unlike for immich as a whole

**Pull request merged**